pkg/identity/cache/local.go

```diff
@@ -0,0 +1,189 @@
+func (l *localIdentityCache) allocateNumericIdentity() (identity.NumericIdentity, error) {
+	firstID := l.nextNumericIdentity
+	for {
+		if _, ok := l.identitiesByID[l.nextNumericIdentity]; !ok {
+			id := l.nextNumericIdentity
+			l.bumpNextNumericIdentity()
+			return id, nil
+		}
+
+		l.bumpNextNumericIdentity()
+		if l.nextNumericIdentity == firstID {
+			return 0, fmt.Errorf("out of local identity space")
+		}
+	}
+}
+
+// lookupOrCreate searches for the existence of a local identity with the given
+// labels. If it exists, the reference count is incremented and the identity is
+// returned. If it does not exist, a new identity is created with a unique
+// numeric identity. All identities returned by lookupOrCreate() must be
+// released again via localIdentityCache.release().
+func (l *localIdentityCache) lookupOrCreate(lbls labels.Labels) (*identity.Identity, bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	stringRepresentation := string(lbls.SortedList())
+	if id, ok := l.identitiesByLabels[stringRepresentation]; ok {
+		id.ReferenceCount++
+		return id, false, nil
+	}
+
+	numericIdentity, err := l.allocateNumericIdentity()
+	if err != nil {
+		return nil, false, err
+	}
+
+	id := &identity.Identity{
+		ID:             numericIdentity | identity.LocalIdentityFlag,
+		Labels:         lbls,
+		LabelArray:     lbls.LabelArray(),
+		ReferenceCount: 1,
+	}
+
+	l.identitiesByLabels[stringRepresentation] = id
+	l.identitiesByID[numericIdentity] = id
+
+	if l.events != nil {
+		l.events <- allocator.AllocatorEvent{
+			Typ: kvstore.EventTypeCreate,
+		}
+	}
+
+	return id, true, nil
+}
+
+// release releases a local identity from the cache. true is returned when the
+// last use of the identity has been released and the identity has been
+// forgotten.
+func (l *localIdentityCache) release(id *identity.Identity) bool {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	stringRepresentation := string(id.Labels.SortedList())
+	if id, ok := l.identitiesByLabels[stringRepresentation]; ok {
+		switch {
+		case id.ReferenceCount > 1:
+			id.ReferenceCount--
+			return false
+
+		case id.ReferenceCount == 1:
+			// Release is only attempted once, when the reference count is
+			// hitting the last use
+			stringRepresentation := string(id.Labels.SortedList())
+			delete(l.identitiesByLabels, stringRepresentation)
+			delete(l.identitiesByID, (id.ID & ^identity.LocalIdentityFlag))
+
+			if l.events != nil {
+				l.events <- allocator.AllocatorEvent{
+					Typ: kvstore.EventTypeDelete,
+				}
+			}
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// lookup searches for a local identity matching the given labels and returns
+// it. If found, the reference count is NOT incremented and thus release must
+// NOT be called.
+func (l *localIdentityCache) lookup(lbls labels.Labels) *identity.Identity {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+
+	stringRepresentation := string(lbls.SortedList())
```

Comment:
This may be outside the scope of this PR, but we sort labels a lot in these functions. Is there any way to ensure that the labels are sorted further up the call-stack? This could prove to be an unreasonable expectation on my side.

Reply:
I think we can cache this but I agree, outside of the scope of this PR.